testing/base/source_gen_test.cpp

```diff
@@ -0,0 +1,152 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "testing/base/source_gen.h"
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include "common/set.h"
+#include "testing/base/gtest_main.h"
+#include "toolchain/driver/driver.h"
+
+namespace Carbon::Testing {
+namespace {
+
+using ::testing::AllOf;
+using ::testing::ContainerEq;
+using ::testing::Contains;
+using ::testing::Each;
+using ::testing::Eq;
+using ::testing::Ge;
+using ::testing::Le;
+using ::testing::MatchesRegex;
+using ::testing::SizeIs;
+
+// Tiny helper to sum the sizes of a range of ranges. Uses a template to avoid
+// hard coding any specific types for the two ranges.
+template <typename T>
+static auto SumSizes(const T& range) -> ssize_t {
+  ssize_t sum = 0;
+  for (const auto& inner_range : range) {
+    sum += inner_range.size();
+  }
+  return sum;
+}
+
+TEST(SourceGenTest, Ids) {
+  SourceGen gen;
+
+  auto ids = gen.GetShuffledIds(1000);
+  EXPECT_THAT(ids.size(), Eq(1000));
+  for (llvm::StringRef id : ids) {
+    EXPECT_THAT(id, MatchesRegex("[A-Za-z][A-Za-z0-9_]*"));
+  }
+
+  // Check that repeated calls are different in interesting ways, but have the
+  // exact same total bytes.
+  ssize_t ids_size_sum = SumSizes(ids);
+  for ([[maybe_unused]] int i : llvm::seq(10)) {
+    auto ids2 = gen.GetShuffledIds(1000);
+    EXPECT_THAT(ids2, SizeIs(1000));
+    // Should be (at least) a different shuffle of identifiers.
+    EXPECT_THAT(ids2, Not(ContainerEq(ids)));
+    // But the sum of lengths should be identical.
+    EXPECT_THAT(SumSizes(ids2), Eq(ids_size_sum));
+  }
+
+  // Check length constraints have the desired effect.
+  ids = gen.GetShuffledIds(1000, /*min_length=*/10, /*max_length=*/20);
+  EXPECT_THAT(ids, Each(SizeIs(AllOf(Ge(10), Le(20)))));
+
+  // Check that uniform id length results in exact coverage of each possible
+  // length for an easy case.
+  ids = gen.GetShuffledIds(100, /*min_length=*/10, /*max_length=*/19,
+                           /*uniform=*/true);
+  EXPECT_THAT(ids, Contains(SizeIs(10)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(11)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(12)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(13)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(14)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(15)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(16)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(17)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(18)).Times(10));
+  EXPECT_THAT(ids, Contains(SizeIs(19)).Times(10));
+}
+
+// Largely covered by `Ids`, but need to check for uniqueness specifically.
+TEST(SourceGenTest, UniqueIds) {
```

Comment:
I think it'd be good to test the length distribution and rounding behavior here.

Reply:
Yes, I think I meant a non-unique test :)

One simple thing you could maybe test is that identifiers of length 4 are the most common length, which I think ought to hold for N >= 4 or so (with the adjustment I suggested earlier), and shouldn't be too much of a change detector test since I don't think we'd expect that property to change as we tune the distribution.

Reply:
Tried to add something like this...